packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -206,23 +206,39 @@ export class OdspDocumentService implements IDocumentService {
                 ? Promise.resolve(null)
                 : this.getWebsocketToken!(options);
             const joinSessionPromise = this.joinSession(requestWebsocketTokenFromJoinSession).catch((e) => {
-                let code: string | undefined;
+                const stack: any = [];
                 try {
-                    code = e?.response ? JSON.parse(e?.response)?.error?.code : undefined;
+                    if(e !== undefined) {
+                        stack.push(e.code);
```

Comment:
e is an exception, most likely (but no guarantees) thrown from throwOdspNetworkError.
There is no code property on it. Ideally we define proper interface there (currently it's OdspError, but it's missing a bunch of properties that this methods adds to any error thrown from it). And then use it here such that compiler can help us find bugs like this.

Reply:
from tracking down joinSession, `e` can also be ThrottlingError from `checkForEpochError `. do we need to add properties to that?

Reply:
for the new interface, are we keeping the properties `response` and `error`?

Reply:
Yes, ThrottlingError  thrown in checkForEpochError() should be handled same way as any unknown error, i.e. it will not pass the smell test of exception we expect, and thus it does not have error info this code is looking for.

I'd do something like that:

```typescript
interface IFacetCodes {
   facetCodes?: number[];
}

function parseFacetCodes(response: Response): number[] | undefined {
   ... /// logic to extract it.
}

export function createOdspNetworkError(...): OdspError & LoggingError & IFacetCodes
{
   ...
   const facetInfo: IFacetCodes = { facetCodes: parseFacetCodes(response) };
   error.addTelemetryProperties(facetInfo);
   assert(error.facetCodes === facetInfo.facetCodes); // nice to validate that, as behavior of addTelemetryProperties() may change in the future. We should add UT validating property stays on this object, and with UTs we may remove this assert.
   return error;
}

public async connectToDeltaStream(client: IClient): Promise<IDocumentDeltaConnection> {
   ...
   const joinSessionPromise = this.joinSession(requestWebsocketTokenFromJoinSession).catch((e) => {
      const likelyFacetCodes = e as IFacetCodes;
      if (likelyFacetCodes.facetCodes !== undefined) {
          ...
      }
   }
}

```

Reply:
`response` is not a property for OdspError  and `error` is not a property on `Response`, do they both need to be added?

Reply:
oh nvm I think I get it